sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,672 @@
+
+We solve this problem by changing the post-link tool to bundle kernels
+according to the aspects that they use.
+
+The post-link tool is also a convenient place to issue a diagnostic when a
+function uses aspects that it is not allowed to use (i.e. the function is
+decorated with `[[sycl::requires()]]` and it uses some optional feature that
+corresponds to an aspect that is not in the `[[sycl::requires()]]` list).
+
+The post-link tool achieves this by examining the static call graph of each
+kernel and each exported device function.  When this tool computes the static
+call graph, it considers all code in any of the translation units that are
+being linked together.  This may not be the complete call graph, however, in
+cases where a kernel calls out to a device function that is defined in a
+different shared library.
+
+**NOTE**: In this context, "exported device function" means a device function
+that is exported from a shared library as defined by [Device Code Dynamic
+Linking][5].
+
+[5]: <https://github.com/intel/llvm/blob/sycl/sycl/doc/SharedLibraries.md>
+
+#### Pass to identify aspects used by each device function
+
+This pass operates on the static call graph for each kernel and each exported
+device function, propagating the aspects from the `!intel_used_aspects` and
+`!intel_allowed_aspects` metadata from the leaves of the call graph up to their
+callers.  The result of this pass is that each device function is labeled with
+a *Used* set of aspects which is computed as the union of the following:
+
+* The aspects in the function's `!intel_used_aspects` metadata (if any).
+* The aspects in the function's `!intel_allowed_aspects` metadata (if any).
+* The aspects in the *Used* set of all functions called by this function.
+
+Once the *Used* set of aspects is known for each function, the post-link tool
+compares this set of aspects with the aspects from any `!intel_allowed_aspects`
+metadata.  If the function has this metadata and if the *Used* set contains
+aspects not in that set, it issues a warning indicating that the function uses
+aspects that are not in the `[[sycl::requires()]]` list.  Unfortunately, the
+post-link tool is unable to include the source position of the code that uses
+the aspect in question.  To compensate, the warning message must include
+instructions telling the user how to run the clang static analyzer which
+provides a better diagnostic.  This analysis phase is described in more detail
+below.
+
+#### Changes to the device code split algorithm
+
+The algorithm for splitting device functions into images must be changed to
+account for the *Used* aspects of each kernel or exported device function.  The
+goal is to ensure that two kernels or exported device functions are only
+bundled together into the same device image if their *Used* sets are identical.
+
+We must also split two kernels into different device images if they have
+different `[[sycl::reqd_sub_group_size()]]` or different
```

Comment:
Note: our SubGroups extension supports [`named_sub_group_size` attribute](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/SubGroup/SYCL_INTEL_sub_group.asciidoc#attributes) which should probably be taken into account here as well

Reply:
Thanks for pointing this out.  I know about that DPC++ extension, but I don't think I've seen a design document for it.  I presume this will rely on an underlying SPIR-V extension?  Do you know if there is a proposed document for that SPIR-V extension?  The details of the SPIR-V extension might influence how the kernel bundling should work.

It seems like there are two options for the bundling which might make sense:

1. Allow kernels with sub-group size `primary` to be bundled together with kernels that have no required sub-group size.  (This makes sense because `primary` is the default.)  Or,

2. Allow kernels with sub-group size `automatic` and `primary` to be bundled together with each other and with kernels that have no required sub-group size.

Option (1) is clearly safer, but I can't think of any reason why option (2) would cause problems.

Tagging @Pennycook since he is the author of that DPC++ extension.

Reply:
I changed the design document to handle this case in a818156a5c94c178791334b76bca83bb59d70e56.  I think my option (2) is safe, so that's what I chose.